internal/internal_task_handlers.go

```diff
@@ -1866,6 +1861,27 @@ func (ath *activityTaskHandlerImpl) getRegisteredActivityNames() (activityNames
 	return
 }
 
+func (ath *activityTaskHandlerImpl) getHeartbeatThrottleInterval(heartbeatTimeout time.Duration) time.Duration {
+	// Build the heartbeat throttle interval as 80% of timeout if timeout present
+	// or use default if not
+	var heartbeatThrottleInterval time.Duration
+	if heartbeatTimeout > 0 {
+		heartbeatThrottleInterval = time.Duration(0.8 * float64(heartbeatTimeout))
+	} else if ath.defaultHeartbeatThrottleInterval > 0 {
+		heartbeatThrottleInterval = ath.defaultHeartbeatThrottleInterval
+	} else {
+		heartbeatThrottleInterval = defaultDefaultHeartbeatThrottleInterval
+	}
+	// Limit interval to a max
+	if ath.maxHeartbeatThrottleInterval > 0 && heartbeatThrottleInterval > ath.maxHeartbeatThrottleInterval {
+		heartbeatThrottleInterval = ath.maxHeartbeatThrottleInterval
+	} else if ath.maxHeartbeatThrottleInterval == 0 && heartbeatThrottleInterval > defaultMaxHeartbeatThrottleInterval {
+		heartbeatThrottleInterval = defaultMaxHeartbeatThrottleInterval
+	}
```

Comment:
I find this a bit complicated to read.

It would have been easier to only have to consider `ActivityOptions.heartbeatTimeout`, `WorkerOptions.{default,max}HeartbeatThrottleInterval` and have defaults for those options provided outside of this method.

Reply:
They are provided outside this method too. But there are a couple of ways this is called in test code, so I wanted to do it here just to be safe.

Reply:
So I'd split it into 2 methods, one that sets the default options and one that does the calculation to reduce the complexity.

Reply:
Those methods would be a few lines and each called only once.

Reply:
Can we do something like this?

In `func newActivityTaskHandlerWithCustomProvider`
do:
```go
defaultHeartbeatThrottleInterval:
  providedOrDefaultInterval(params.DefaultHeartbeatThrottleInterval, defaultDefaultHeartbeatThrottleInterval),
maxHeartbeatThrottleInterval:
  providedOrDefaultInterval(params.MaxHeartbeatThrottleInterval, defaultMaxHeartbeatThrottleInterval),
```

And make `getHeartbeatThrottleInterval` something like:
```go
if heartbeatTimeout > 0 {
		heartbeatThrottleInterval = time.Duration(0.8 * float64(heartbeatTimeout))
} else {
		heartbeatThrottleInterval = ath.defaultHeartbeatThrottleInterval
}

if heartbeatThrottleInterval > ath.maxHeartbeatThrottleInterval {
		heartbeatThrottleInterval = ath.maxHeartbeatThrottleInterval
}
return heartbeatThrottleInterval
```

If we're missing the initialization in tests just make sure ActivityTaskHandler is properly initialized.